src/som/interpreter/nodes/ResolvingImplicitReceiverSend.java

```diff
@@ -73,38 +73,49 @@ public Object doPreEvaluated(final VirtualFrame frame, final Object[] args) {
     Lock lock = getLock();
     try {
       lock.lock();
-      newNode = specialize(args);
+      newNode = specialize(frame, args);
     } finally {
       lock.unlock();
     }
     return newNode.doPreEvaluated(frame, args);
   }
 
-  private PreevaluatedExpression specialize(final Object[] args) {
+  private PreevaluatedExpression specialize(final VirtualFrame frame, final Object[] args) {
     if (replacedBy != null) {
       // already has been specialized
       if (newReceiverNodeForOuterSend != null) {
         // need to recalculate the real receiver for outer sends
-        args[0] = newReceiverNodeForOuterSend.executeEvaluated(args[0]);
+        args[0] = newReceiverNodeForOuterSend.executeGeneric(frame);
       }
       return replacedBy;
     }
     // first check whether it is an outer send
     // it it is, we get the context level of the outer send and rewrite to one
-    MixinIdAndContextLevel result = currentScope.lookupSlotOrClass(selector);
-    if (result != null && result.contextLevel > 0) {
-      newReceiverNodeForOuterSend = OuterObjectReadNodeGen.create(
-          result.contextLevel, mixinId, result.mixinId,
-          argumentNodes[0]).initialize(sourceSection);
+    List<MixinDefinitionId> result = currentScope.lookupSlotOrClass(selector);
+    if (result != null && result.size() > 1) {
+      assert mixinId == result.get(result.size() - 1);
+
       ExpressionNode[] msgArgNodes = argumentNodes.clone();
-      msgArgNodes[0] = newReceiverNodeForOuterSend;
+      MixinDefinitionId currentMixin = mixinId;
+      MixinDefinitionId enclosingMixin;
+
+      for (int i = result.size() - 2; i > -1; i--) {
+        enclosingMixin = result.get(i);
+
+        newReceiverNodeForOuterSend =
```

Comment:
This update is problematic, because it is a transitory update, which is not leading to a correct lookup, but is visible to other threads. Will fix this.